llvm/include/llvm/ADT/SmallVector.h

```diff
@@ -468,7 +469,7 @@ void SmallVectorTemplateBase<T, TriviallyCopyable>::takeAllocationForGrow(
     free(this->begin());
 
   this->BeginX = NewElts;
-  this->Capacity = NewCapacity;
+  this->Capacity = static_cast<SmallVectorSizeType<T>>(NewCapacity);
```

Comment:
Should we add a `set_capacity(size_t)` method to `SmallVectorBase`, to make this cast a bit neater?

Reply:
added. please also take a look on the new function (i'm not a native English speaker so maybe there are errors in the docs. albeit I pretty much copy-pasted the relevant `set_size` wording).

Reply:
Thanks.

After seeing this change, I think it would probably make sense to include this line in the new function:
```
this->BeginX = NewElts;
```
... setting the allocation pointer and the capacity together makes sense.

This would also change the name of the new function:
```
  /// Set the allocation to \p A and the capacity to \p M.
  ///
  /// This does not construct or destroy any elements in the vector.
  /// This does not clean up any existing allocation.
void set_allocation_and_capacity(void *A, size_t M) {
```

Reply:
i'm actually hesitant a bit: the `set_capacity()` has a much broader usage scope than the new function. i guess there are also cases when we only want to set the capacity?

Reply:
> i guess there are also cases when we only want to set the capacity?

Such as?

I think setting a capacity is intrinsically linked to setting a new storage pointer, which is why it makes sense to do both at once..

Reply:
FWIW, I agree with @browneee; these are intrinsically linked; the point of capacity is to understand the storage pointer.

Reply:
> > i guess there are also cases when we only want to set the capacity?
> 
> Such as?

Was thinking of realloc() but we'd anyway reset the pointer (since same memory block being "reused" is an implementation detail).

> I think setting a capacity is intrinsically linked to setting a new storage pointer, which is why it makes sense to do both at once..

yeah, fair point. I was wondering whether `data start != allocation start` but these seem to always be equal (I guess the case when they are not is the custom alignment requirements but I haven't found anything like this).

Reply:
one thing is whether `set_allocation_and_capacity` is a good name. to me, an "allocation" would mean both the start and the capacity (or end). so named the function `set_allocation_range`. There's however already `replaceAllocation()` that accepts both so maybe `set_allocation` is enough? anyways, guessing it's for you guys to decide what makes more sense here since you seem to interact with this code more.

Reply:
I prefer `set_allocation_and_capacity`, but I think any of these options are fine.